stan/math/prim/err/check_finite.hpp

```diff
@@ -1,87 +1,27 @@
 #ifndef STAN_MATH_PRIM_ERR_CHECK_FINITE_HPP
 #define STAN_MATH_PRIM_ERR_CHECK_FINITE_HPP
 
-#include <stan/math/prim/meta.hpp>
-#include <stan/math/prim/err/is_scal_finite.hpp>
-#include <stan/math/prim/err/throw_domain_error.hpp>
-#include <stan/math/prim/err/throw_domain_error_vec.hpp>
-#include <stan/math/prim/fun/Eigen.hpp>
-#include <stan/math/prim/fun/get.hpp>
-#include <stan/math/prim/fun/size.hpp>
-#include <stan/math/prim/fun/value_of.hpp>
-#include <stan/math/prim/fun/value_of_rec.hpp>
-#include <cmath>
+#include <stan/math/prim/err/elementwise_check.hpp>
 
 namespace stan {
 namespace math {
 
-namespace internal {
-template <typename T_y, bool is_vec>
-struct finite {
-  static void check(const char* function, const char* name, const T_y& y) {
-    if (!is_scal_finite(y)) {
-      throw_domain_error(function, name, y, "is ", ", but must be finite!");
-    }
-  }
-};
-
-template <typename T_y>
-struct finite<T_y, true> {
-  static void check(const char* function, const char* name, const T_y& y) {
-    for (size_t n = 0; n < stan::math::size(y); n++) {
-      if (!is_scal_finite(stan::get(y, n))) {
-        throw_domain_error_vec(function, name, y, n, "is ",
-                               ", but must be finite!");
-      }
-    }
-  }
-};
-}  // namespace internal
-
 /**
  * Check if <code>y</code> is finite.
  * This function is vectorized and will check each element of
  * <code>y</code>.
- * @tparam T_y Type of y
- * @param function Function name (for error messages)
- * @param name Variable name (for error messages)
- * @param y Variable to check
+ * @tparam T_y type of y
+ * @param function function name (for error messages)
+ * @param name variable name (for error messages)
+ * @param y variable to check
  * @throw <code>domain_error</code> if y is infinity, -infinity, or NaN
  */
 template <typename T_y>
 inline void check_finite(const char* function, const char* name, const T_y& y) {
-  internal::finite<T_y, is_vector_like<T_y>::value>::check(function, name, y);
+  auto is_good = [](const auto& y) { return std::isfinite(y); };
+  elementwise_check(is_good, function, name, y, ", but must be finite!");
 }
 
-/**
- * Return <code>true</code> is the specified matrix is finite.
- *
- * @tparam T type of elements in the matrix
- * @tparam R number of rows, can be Eigen::Dynamic
- * @tparam C number of columns, can be Eigen::Dynamic
- *
- * @param function name of function (for error messages)
- * @param name variable name (for error messages)
- * @param y matrix to test
- * @return <code>true</code> if the matrix is finite
- **/
-namespace internal {
-template <typename T, int R, int C>
-struct finite<Eigen::Matrix<T, R, C>, true> {
-  static void check(const char* function, const char* name,
-                    const Eigen::Matrix<T, R, C>& y) {
-    if (!value_of(y).allFinite()) {
```

Comment:
We want this allFinite call I think. (It does need to be value_of_rec(y).) This gives Eigen a chance to do something clever. See also: https://github.com/stan-dev/math/pull/1005. I'm not really sure what to do here?

Reply:
@peterwicksstringfield, are you suggesting this should be specialized for Eigen types? That would make sense to me. Especially if we could have a microbenchmark showing the difference in performance.